Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1814 @@
+{
+  NSUInteger	c = [otherArray count];
+
+  if (c > 0)
+    {
+      NSUInteger	i;
+      IMP	get = [otherArray methodForSelector: oaiSel];
+      IMP	rem = [self methodForSelector: @selector(removeObject:)];
+
+      for (i = 0; i < c; i++)
+	(*rem)(self, @selector(removeObject:), (*get)(otherArray, oaiSel, i));
+    }
+}
+
+- (void)removeObjectsInRange:(NSRange)aRange
+{
+  NSUInteger	i;
+  NSUInteger	s = aRange.location;
+  NSUInteger	c = [self count];
+
+  i = aRange.location + aRange.length;
+
+  if (c < i)
+    i = c;
+
+  if (i > s)
+    {
+      IMP	rem = [self methodForSelector: remSel];
+
+      while (i-- > s)
+	{
+	  (*rem)(self, remSel, i);
+	}
+    }
+}
+
+- (void)removeAllObjects
+{
+  NSUInteger	c = [self count];
+
+  if (c > 0)
+    {
+      IMP	remLast = [self methodForSelector: rlSel];
+
+      while (c--)
+	{
+	  (*remLast)(self, rlSel);
+	}
+    }  
+}
+
+- (void)replaceObjectAtIndex:(NSUInteger)index // required override...
+                  withObject:(id)object
+{
+  [self subclassResponsibility: _cmd];
+}
+
+- (void) replaceObjectsAtIndexes: (NSIndexSet *)indexes
+                     withObjects: (NSArray *)objects
+{
+  NSUInteger count = [indexes count];
+  NSUInteger indexArray[count];
+
+  [indexes getIndexes: indexArray
+             maxCount: count
+         inIndexRange: NULL];
+
+  [self _removeObjectsFromIndices: indexArray
+		       numIndices: count]; 
+}
+
+- (void) replaceObjectsInRange: (NSRange)aRange
+                   withObjects: (const id[])objects
+                         count: (NSUInteger)count
+{
+  id o = nil;
+  NSUInteger i = 0;
+  
+  if (count < (aRange.location + aRange.length))
+    [NSException raise: NSRangeException
+		 format: @"Replacing objects beyond end of const[] id."];
+  [self removeObjectsInRange: aRange];
+
+  for(i = 0; i < count; i++)
+    {
+      o = objects[i];
+      [self insertObject: o atIndex: aRange.location]; 
+    }
+}
+
+- (void)setObject:(id)anObject atIndex:(NSUInteger)anIndex
+{
+  if ([self count] == anIndex)
+    {
+      [self addObject: anObject];
+    }
+  else
+    {
+      [self replaceObjectAtIndex: anIndex withObject: anObject];
+    }
+}
+
+- (void)moveObjectsAtIndexes:(NSIndexSet *)indexes toIndex:(NSUInteger)index
+{
+  NSUInteger i = 0;
+  NSUInteger count = [indexes count];
+  NSUInteger indexArray[count];
+  NSMutableArray *tmpArray = [NSMutableArray arrayWithCapacity: count];
+  id o = nil;
+  NSEnumerator *e = nil;
+  
+  [indexes getIndexes: indexArray
+             maxCount: count
+         inIndexRange: NULL];
+
+  // Build the temporary array....
+  for(i = 0; i < count; i++)
+    {
+      NSUInteger index = indexArray[i];
+      id obj = [self objectAtIndex: index];
+      [tmpArray addObject: obj];
+    }
+
+  // Move the objects
+  e = [tmpArray objectEnumerator];
+  while((o = [e nextObject]) != nil)
+    {
+      [self insertObject: o atIndex: index];
+    }
+
+  // Remove the originals...  
+  for(i = 0; i < count; i++)
```

Comment:
This bit here is definitely wrong. But the rest also needs some testing with the new objects being inserted before and after the old ones.